ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt

```diff
@@ -200,32 +244,92 @@ internal class PaywallViewModelImpl(
         }
     }
 
-    private fun purchasePackage(activity: Activity, packageToPurchase: Package) {
-        if (verifyNoActionInProgressOrStartAction()) { return }
-
-        viewModelScope.launch {
-            try {
-                listener?.onPurchaseStarted(packageToPurchase)
-                val purchaseResult = purchases.awaitPurchase(
-                    PurchaseParams.Builder(activity, packageToPurchase),
-                )
-                listener?.onPurchaseCompleted(purchaseResult.customerInfo, purchaseResult.storeTransaction)
-                Logger.d("Dismissing paywall after purchase")
-                options.dismissRequest()
-            } catch (e: PurchasesException) {
-                if (e.code == PurchasesErrorCode.PurchaseCancelledError) {
-                    trackPaywallCancel()
-                    listener?.onPurchaseCancelled()
+    private suspend fun handlePackagePurchase(activity: Activity) {
+        when (val currentState = _state.value) {
+            is PaywallState.Loaded -> {
+                val selectedPackage = currentState.selectedPackage.value
+                if (!selectedPackage.currentlySubscribed) {
+                    performPurchase(activity, selectedPackage.rcPackage)
                 } else {
-                    listener?.onPurchaseError(e.error)
-                    _actionError.value = e.error
+                    Logger.d("Ignoring purchase request for already subscribed package")
                 }
             }
 
-            finishAction()
+            else -> {
+                Logger.e("Unexpected state trying to purchase package: $currentState")
+            }
         }
     }
 
+    @Suppress("LongMethod")
+    private suspend fun performPurchase(activity: Activity, packageToPurchase: Package) {
+        try {
+            listener?.onPurchaseStarted(packageToPurchase)
+
+            val customPurchaseHandler = myAppPurchaseLogic?.performPurchase
+
+            when (purchases.purchasesAreCompletedBy) {
+                PurchasesAreCompletedBy.MY_APP -> {
+                    checkNotNull(customPurchaseHandler) {
+                        "myAppPurchaseLogic must not be null when purchases.purchasesAreCompletedBy " +
+                            "is PurchasesAreCompletedBy.MY_APP"
+                    }
+                    when (val result = customPurchaseHandler.invoke(activity, packageToPurchase)) {
```

Comment:
Same here, `invoke` can be omitted: 
```suggestion
                    when (val result = customPurchaseHandler(activity, packageToPurchase)) {
```